### ISSUE TITLE
Loading freeze for object-script that tries to access the object list

### ISSUE THREAD

**Issue by Sailsman63** [ArtOfIllusion/ArtOfIllusion#280]
#215 

Introduced a freeze when trying to open certain AOI files. The file in question has an object script that directly modifies a material parameter. (IoR in this case)

The file will load correctly when re-saved without the object script.

**Comment by peteihis** [comment]
I'd suggest, the old script does not have the language set? 

Without language selection the file may even not be successfully saved or at least it defaults to Groovy which causes runtime error on a BeanShell script.

**Comment by Sailsman63** [comment]
Thanks for the thought. Not that simple, though. (I did test, just to be sure.)

The script I'm working with is this one:

```
// courtesy Julian MacDonald - header stripped for readability

scene = script.getScene();
objectName="prism";   // make sure your object has this name (or change the name in quotes to the name of your object)
//
if (scene.getObject(objectName)==null) return;  // return without annoying error if there isn't an object of that name
  else
{
  selObjInfo=scene.getObject(objectName);
  selObj=selObjInfo.object;
  Mat=selObj.getMaterial();
  MatMapping=selObj.getMaterialMapping();
//
//
  IoRVal = script.getParameter("IoRVal");
  Mat.setIndexOfRefraction(IoRVal);
  selObjInfo.setMaterial(Mat,MatMapping);
}
```

When I enter it by hand, I get an alert for a null pointer exception, even if I comment out everything but the `script.getScene()` line. I've also found that a ScriptedObject that does nothing but throw a new NPE can freeze the interface when a file is loaded.

This means we have two things to fix:

- Why is `script.getScene()` returning null?
- Harden scene-opening against NPEs and other exceptions from object scripts

**Comment by Sailsman63** [comment]
And it looks like I was wrong. This has nothing to do with an NPE (though I will want to check into that later)

This particular script reaches into the scene through `scene.getObject(String name)`. This method iterates through the entire collection of objects, looking for a match. However, it gets blocked by the iterator that is already working through the collection of objects.

Not yet clear how to fix this - I could migrate to ArrayList, but then we'd have concurrent modification problems with any code that is adding objects/textures/etc to the list.

Note that the old index-based for-loops were probably unsound as well, they just got lucky.

**Comment by peteihis** [comment]
The same script...

The error at opening an old file was due to the missing language selection. The sript would actually work just as well either as BSh or Groovy. I seem to remember that I got an NPE once or twice.Don't remember if it had any other effects. This morning that did not happen. 

The `getScene().getObject(String nam)` returns null if an object by that name does not exist. And if you try to ´getObject()`on that one you naturally get serious complaints about NPE. 

> it gets blocked by the iterator that is already working through the collection of objects.

Now the queston "What iterator?" (is already there). I haven't had a look at the code but should't these processes be independent of eachothers? Sounds strange anyway.
